File: api/editor/api-editor/src/main/java/ca/bc/gov/tno/areas/reports/controllers/CBRAController.java
package ca.bc.gov.tno.areas.reports.controllers;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.time.LocalDateTime;
import java.time.ZoneId;

import javax.annotation.security.RolesAllowed;
import javax.servlet.http.HttpServletResponse;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.InputStreamResource;
import org.springframework.core.io.Resource;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import ca.bc.gov.tno.core.excel.CBRAReport;
import ca.bc.gov.tno.dal.db.services.interfaces.IActionService;
import ca.bc.gov.tno.dal.db.services.interfaces.IContentService;
import ca.bc.gov.tno.dal.db.services.interfaces.IUserService;

/**
 * ActionController class, provides endpoints for actions.
 */
@RolesAllowed({ "administrator", "editor" })
@RestController("ReportsCBRAController")
@RequestMapping({ "/reports/cbra", "/api/reports/cbra" })
public class CBRAController {

  /**
   * DAL for action.
   */
  private final IContentService contentService;
  private final IUserService userService;
  private final IActionService actionService;

  /**
   * Creates a new instance of a ActionController object, initializes with
   * specified parameters.
   *
   * @param contentService Content service.
   * @param userService    User service.
   * @param actionService  Action service.
   */
  @Autowired
  public CBRAController(final IContentService contentService, final IUserService userService,
      final IActionService actionService) {
    this.contentService = contentService;
    this.userService = userService;
    this.actionService = actionService;
  }

  /**
   * Request a list of all actions from the db.
   *
   * @return
   * @throws IOException
   */
  @PostMapping(path = { "", "/" }, consumes = {
      MediaType.ALL_VALUE }, produces = MediaType.APPLICATION_OCTET_STREAM_VALUE)
  public ResponseEntity<Resource> generateReport(
      @RequestParam(required = true) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime from,
      @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime to,
      HttpServletResponse response) throws IOException {

    if (to == null)
      to = LocalDateTime.now();

    var report = new CBRAReport(contentService, userService, actionService);
    var workbook = report.generateReport(from.atZone(ZoneId.systemDefault()), to.atZone(ZoneId.systemDefault()));

    var outputStream = new ByteArrayOutputStream();
    workbook.write(outputStream);
    workbook.close();

    var inputStream = new ByteArrayInputStream(outputStream.toByteArray());
    var resource = new InputStreamResource(inputStream);

    var contentType = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet";
    var headers = new HttpHeaders();
    headers.setContentType(MediaType.parseMediaType(contentType));
    headers.setContentDispositionFormData("attachment", "cbra.xlsx");

    return ResponseEntity.ok()
        .headers(headers)
        .body(resource);
  }
}

File: api/net/Areas/Reports/Controllers/CBRAController.cs
using System.Net;
using Microsoft.AspNetCore.Mvc;
using Swashbuckle.AspNetCore.Annotations;
using TNO.API.Models;
using TNO.DAL.Services;
using TNO.Keycloak;
using TNO.Reports;

namespace TNO.API.Areas.Reports.Controllers;

/// <summary>
/// CBRAController class, provides CBRA endpoints for the api.
/// </summary>
[ClientRoleAuthorize(ClientRole.Administrator)]
[ApiController]
[Area("reports")]
[ApiVersion("1.0")]
[Route("api/v{version:apiVersion}/[area]/cbra")]
[Route("api/[area]/cbra")]
[Route("v{version:apiVersion}/[area]/cbra")]
[Route("[area]/cbra")]
[ProducesResponseType(typeof(ErrorResponseModel), (int)HttpStatusCode.Unauthorized)]
[ProducesResponseType(typeof(ErrorResponseModel), (int)HttpStatusCode.Forbidden)]
public class CBRAController : ControllerBase
{
    #region Variables
    private readonly ITimeTrackingService _timeTrackingService;
    #endregion

    #region Constructors
    /// <summary>
    /// Creates a new instance of a CBRAController object, initializes with specified parameters.
    /// </summary>
    /// <param name="timeTrackingService"></param>
    public CBRAController(ITimeTrackingService timeTrackingService)
    {
        _timeTrackingService = timeTrackingService;
    }
    #endregion

    #region Endpoints
    /// <summary>
    /// Generate an Excel document CBRA report.
    /// </summary>
    /// <returns></returns>
    [HttpPost]
    [Produces("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet")]
    [ProducesResponseType((int)HttpStatusCode.OK)]
    [SwaggerOperation(Tags = new[] { "CBRA" })]
    public FileResult Generate([FromQuery] DateTime from, [FromQuery] DateTime? to)
    {
        var helper = new CBRAReport(_timeTrackingService);

        var report = helper.GenerateReport(from, to);

        using var stream = new MemoryStream();
        report.Write(stream);
        var bytes = stream.ToArray();

        return File(bytes, "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet", "cbra.xlsx");
    }
    #endregion
}
